### PR TITLE
ISSUE-297 Webadmin API routes for managing the rate limiting plans users

### PR DESCRIPTION
**Pull request by quantranhong1999**
https://github.com/linagora/james-project-private/issues/297
TODO write more tests, refine code.

### CODE REVIEW

File: tmail-backend/webadmin/webadmin-rate-limit/pom.xml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    <artifactId>webadmin-rate-limit</artifactId>
+    <name>Tmail :: WebAdmin :: Team Mailboxes</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>tmail-rate-limiter-api</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/pom.xml:21]: Move the version to a dependency management.

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+                .statusCode(HttpStatus.NOT_FOUND_404)
+                .type(ErrorResponder.ErrorType.NOT_FOUND)
+                .message(String.format("User %s does not exist", username.asString()))
+                .haltError();
+        }
+    }
+
+    public void planIdPreconditions(RateLimitingPlanId planId) {
+        if (!Throwing.supplier(() -> Boolean.unbox(Mono.from(planRepository.isPlanExisted(planId)).block()))
+            .sneakyThrow().get()) {
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java:185]: Can we get this less verbose?

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    private Username extractUsername(Request request) {
+        return Username.of(request.params(USERNAME_PARAM));
+    }
+
+    private RateLimitingPlanId extractPlanId(Request request) {
+        return new RateLimitingPlanId(UUID.fromString(request.params(USERNAME_PARAM)));
+    }
+
+    public void userPreconditions(Username username) {
+        if (!Throwing.supplier(() -> usersRepository.contains(username)).sneakyThrow().get()) {
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java:174]: Can we get this less verbose?

File: tmail-backend/webadmin/webadmin-rate-limit/src/test/java/com/linagora/tmail/webadmin/RateLimitPlanUsersRoutesTest.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+import com.linagora.tmail.rate.limiter.api.RateLimitingPlanUserRepository;
+import com.linagora.tmail.rate.limiter.api.memory.MemoryRateLimitingPlanUserRepository;
+
+import io.restassured.RestAssured;
+import io.restassured.path.json.JsonPath;
+import reactor.core.publisher.Mono;
+
+public class RateLimitPlanUsersRoutesTest {
+    private static final String BASE_PATH = "/rate-limit/plans";
+    private static final String ADD_USER_TO_PLAN_PATH = "/rate-limit/plans/%s/users/%s";
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/test/java/com/linagora/tmail/webadmin/RateLimitPlanUsersRoutesTest.java:57]: Users is the base resource here that rate limiting plans are attached to.

THus it should come first in the URL.

`/users/bob@domain.tld/rate-limiting-plan` to get the plan of a user for instance.

`/rate-limit-plans/%s/users` is OK to list the users attached to a plan (as plan is here the base resource)

**Comment by vttranlina** [comment]
https://github.com/linagora/tmail-backend/blob/6cd46923a91cee229208c89077520256ecff724c/tmail-backend/rate-limiter/rate-limiter-api/pom.xml#L14

replace
 <packaging>pom</packaging>
-> 
 <packaging>jar</packaging>

my mistake, pls rebase lastest my pr https://github.com/linagora/tmail-backend/pull/333

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    @Override
+    public String getBasePath() {
+        return RATE_LIMIT_BASE_PATH;
+    }
+
+    @Override
+    public void define(Service service) {
+        service.post(ATTACH_PLAN_TO_USER_PATH, attachPlanToUser(), jsonTransformer);
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java:81]: IMO should more be a PUT

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            userPreconditions(username);
+            return Mono.from(planUserRepository.getPlanByUser(username))
+                .map(planId -> new RateLimitingPlanIdResponse(planId.value().toString()))
+                .onErrorResume(NullPointerException.class, e -> {
+                    throw ErrorResponder.builder()
+                        .statusCode(HttpStatus.BAD_REQUEST_400)
+                        .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                        .message(e.getMessage())
+                        .haltError();
+                })
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java:136]: I would enjoy that the code quality to be good enough for us not to need to catch NPEs at the presentation layer!

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            userPreconditions(username);
+            return Mono.from(planUserRepository.revokePlan(username))
+                .then(Mono.just(Responses.returnNoContent(response)))
+                .onErrorResume(NullPointerException.class, e -> {
+                    throw ErrorResponder.builder()
+                        .statusCode(HttpStatus.BAD_REQUEST_400)
+                        .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                        .message(e.getMessage())
+                        .haltError();
+                })
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java:160]: Idem! Why are we throwing NPEs in the first place? We should not!

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.NOT_FOUND_404)
+                .type(ErrorResponder.ErrorType.NOT_FOUND)
+                .message(String.format("User %s does not exist", username.asString()))
+                .haltError();
+        }
+    }
+
+    public void planIdPreconditions(RateLimitingPlanId planId) {
+        if (Boolean.FALSE.equals(Mono.from(planRepository.planExists(planId)).block())) {
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java:184]:
```suggestion
        if (Mono.from(planRepository.planExists(planId)).block()) {
```

Is easier for the mere mortal I am. Our target audience is not the intelliJ compiler.

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+    private RateLimitingPlanId extractPlanId(Request request) {
+        return new RateLimitingPlanId(UUID.fromString(request.params(PLAN_ID_PARAM)));
+    }
+
+    public void userPreconditions(Username username) throws UsersRepositoryException {
+        if (!usersRepository.contains(username)) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.NOT_FOUND_404)
+                .type(ErrorResponder.ErrorType.NOT_FOUND)
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java:177]: IMO static imports would help.

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+                .block();
+        };
+    }
+
+    private Username extractUsername(Request request) {
+        return Username.of(request.params(USERNAME_PARAM));
+    }
+
+    private RateLimitingPlanId extractPlanId(Request request) {
+        return new RateLimitingPlanId(UUID.fromString(request.params(PLAN_ID_PARAM)));
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java:170]:
```suggestion
        return RateLimitingPlanId.parse(request.params(PLAN_ID_PARAM));
```

Encapsulation issue. IMO we should not leak the underlying id format and add a `parse` helper method for doing just that!

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+            return Flux.from(planUserRepository.listUsers(planId))
+                .map(Username::asString)
+                .collectList()
+                .onErrorResume(NullPointerException.class, e -> {
+                    throw ErrorResponder.builder()
+                        .statusCode(HttpStatus.BAD_REQUEST_400)
+                        .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                        .message(e.getMessage())
+                        .haltError();
+                })
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java:119]: NPE spotted.

In good code NPEs are not supposed to happen cause we do not use null!

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+            planIdPreconditions(planId);
+            return Mono.from(planUserRepository.applyPlan(username, planId))
+                .then(Mono.just(Responses.returnNoContent(response)))
+                .onErrorResume(NullPointerException.class, e -> {
+                    throw ErrorResponder.builder()
+                        .statusCode(HttpStatus.BAD_REQUEST_400)
+                        .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                        .message(e.getMessage())
+                        .haltError();
+                })
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java:101]: NPE...

File: tmail-backend/webadmin/webadmin-rate-limit/src/test/java/com/linagora/tmail/webadmin/RateLimitPlanUsersRoutesTest.java [hunk 13]
```diff
@@ -0,0 +1,452 @@
...
+                .body()
+                .asString();
+
+            assertThat(response).isEmpty();
+            assertThatThrownBy(() -> Mono.from(planUserRepository.getPlanByUser(BOB)).block())
+                .isInstanceOf(RateLimitationPlanNotFoundException.class);
+        }
+
+        @Test
+        void shouldBeIndent() throws UsersRepositoryException {
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/test/java/com/linagora/tmail/webadmin/RateLimitPlanUsersRoutesTest.java:422]: shouldBeIdempotent ?

**Review by chibenwa — APPROVED** [review]

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java [hunk 14]
```diff
@@ -92,13 +90,6 @@ private Route attachPlanToUser() {
             planIdPreconditions(planId);
             return Mono.from(planUserRepository.applyPlan(username, planId))
                 .then(Mono.just(Responses.returnNoContent(response)))
-                .onErrorResume(NullPointerException.class, e -> {
-                    throw ErrorResponder.builder()
-                        .statusCode(HttpStatus.BAD_REQUEST_400)
-                        .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
-                        .message(e.getMessage())
-                        .haltError();
-                })
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java]: I'am just curious.... Why did you wrote this in the first place?

It looks like dead code that was never called....

Comment (quantranhong1999) [reply tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java]: In repository API we have a Precondition check username/planId null. I was not sure the code parsing these in James handles all exception cases and maybe returns null.
I think this is overthinking...

**Review by Arsnael — APPROVED** [review]

File: tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java (hunk 14 above)
Comment (chibenwa) [reply tmail-backend/webadmin/webadmin-rate-limit/src/main/java/com/linagora/tmail/webadmin/RateLimitPlanUserRoutes.java]: TDD is the solution to overthinking and to writing the code you truely need,

**Comment by chibenwa** [comment]
Rebase ?

**Comment by Arsnael** [comment]
@quantranhong1999 Can you rebase and solve conflicts?

**Pull request merged**